pennylane/qnode.py

```diff
@@ -461,9 +467,26 @@ def evaluate(self, args, **kwargs):
         Returns:
             float, array[float]: output expectation value(s)
         """
-        if not self.ops:
-            # construct the circuit
-            self.construct(args, **kwargs)
+        if not self.ops or self.eager:
+            if self.num_variables is not None:
+                # circuit construction has previously been called
+                if len(list(_flatten(args))) == self.num_variables:
+                    # only construct the circuit if the number
+                    # of arguments matches the allowed number
+                    # of variables.
+                    # This avoids construction happening
+                    # via self._pd_analytic, where temporary
+                    # variables are appended to the argument list.
+
+                    # flatten and unflatten arguments
+                    flat_args = list(_flatten(args))
+                    shaped_args = unflatten(flat_args, self.args_shape)
+
+                    # construct the circuit
+                    self.construct(shaped_args, **kwargs)
+            else:
+                # construct the circuit
+                self.construct(args, **kwargs)
```

Comment:
For the non-user-facing methods like `construct` and `evaluate`, I would not use the ** operator for kwargs, much like we don't use the * operator for args. Instead we treat args as a tuple, and kwargs as a dict.

Reply:
Fixed